server.go

```diff
@@ -0,0 +1,273 @@
+			Summary: "Domain has already been submitted",
+			Message: fmt.Sprintf("Domain is already pending. It was submitted on %s.", formattedDate),
+		}
+		issues = hstspreload.Issues{
+			Errors:   issues.Errors,
+			Warnings: append(issues.Warnings, issue),
+		}
+	case StatusPreloaded:
+		issue := hstspreload.Issue{
+			Code:    "server.preload.already_preloaded",
+			Summary: "Domain is already preloaded",
+			Message: "The domain is already preloaded.",
+		}
+		issues = hstspreload.Issues{
+			Errors:   append(issues.Errors, issue),
+			Warnings: issues.Warnings,
+		}
+	default:
+		issue := hstspreload.Issue{
+			Code:    "internal.server.preload.unknown_status",
+			Summary: "Internal error",
+			Message: "Cannot preload; could not find domain status.",
+		}
+		issues = hstspreload.Issues{
+			Errors:   append(issues.Warnings, issue),
+			Warnings: issues.Warnings,
+		}
+	}
+
+	writeJSONOrBust(w, issues)
+}
+
+func pending(w http.ResponseWriter, r *http.Request) {
+	names, err := domainsWithStatus(StatusPending)
+	if err != nil {
+		msg := fmt.Sprintf("Internal error: could not retrieve pending list. (%s)\n", err)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	fmt.Fprintf(w, "[\n")
+	for i, name := range names {
+		comma := ","
+		if i+1 == len(names) {
+			comma = ""
+		}
+
+		fmt.Fprintf(w, `    { "name": "%s", "include_subdomains": true, "mode": "force-https" }%s
+`, name, comma)
+	}
+	fmt.Fprintf(w, "]\n")
+}
+
+func difference(from []chromiumpreload.Domain, take []chromiumpreload.Domain) (diff []chromiumpreload.Domain) {
+	takeSet := make(map[chromiumpreload.Domain]bool)
+	for _, elem := range take {
+		takeSet[elem] = true
+	}
+
+	for _, elem := range from {
+		if !takeSet[elem] {
+			diff = append(diff, elem)
+		}
+	}
+
+	return diff
+}
+
+func update(w http.ResponseWriter, r *http.Request) {
+	// Get preload list.
+	preloadList, listErr := chromiumpreload.GetLatest()
+	if listErr != nil {
+		msg := fmt.Sprintf(
+			"Internal error: could not retrieve latest preload list. (%s)\n",
+			listErr,
+		)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+	var actualPreload []chromiumpreload.Domain
+	for _, entry := range preloadList.Entries {
+		if entry.Mode == chromiumpreload.ForceHTTPS {
+			actualPreload = append(actualPreload, entry.Name)
+		}
+	}
+
+	// Get domains currently recorded as preloaded.
+	databasePreload, dbErr := domainsWithStatus(StatusPreloaded)
+	if dbErr != nil {
+		msg := fmt.Sprintf(
+			"Internal error: could not retrieve domain names previously marked as preloaded. (%s)\n",
+			dbErr,
+		)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	// Calculate values that are out of date.
+	var updates []DomainState
+
+	added := difference(actualPreload, databasePreload)
+	for _, name := range added {
+		updates = append(updates, DomainState{
+			Name:   name,
+			Status: StatusPreloaded,
+		})
+	}
+
+	removed := difference(databasePreload, actualPreload)
```

Comment:
What happens if the same domain/entry is added and removed at the same time? Is that even possible to force?